Fix and align ticket age check in ssl_ticket.c for TLS 1.2 and TLS 1.3

ronald-cron-arm opened the pull request:
## Description
Fix and align ticket age check in ssl_ticket.c for TLS 1.2 and TLS 1.3 

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

yuhaoth reviewed (changes requested):
Just take a quick look.

yuhaoth commented on library/ssl_ticket.c:
```diff
@@ -331,7 +338,7 @@ int mbedtls_ssl_ticket_write(void *p_ticket,
     key = &ctx->keys[ctx->active];
 
-    *ticket_lifetime = ctx->ticket_lifetime;
+    *ticket_lifetime = key->lifetime;
```
`session->ticket_lifetime` MUST be update before `mbedtls_ssl_session_save()` . Otherwise, the ticket_lifetime will be different with client received.

ronald-cron-arm replied:
The idea is that we rely solely on ssl_ticket.c to check the ticket lifetime, we do not use the ticket lifetime in `ssl_tls13_server.c`. When we create a ticket we check that its lifetime is less than MBEDTLS_SSL_TLS1_3_MAX_ALLOWED_TICKET_LIFETIME (check not done by ssl_ticket.c). When we use a ticket the parse function of `ssl_ticket.c` does the check of the ticket lifetime and we do not have to do it in `ssl_tls13_server.c`.

yuhaoth replied:
> When we use a ticket the parse function of ssl_ticket.c does the check of the ticket lifetime and we do not have to do it in `ssl_tls13_server.c.`

`ticket_lifetime` is not saved in ticket if we do not set `session->ticket_lifetime` before `mbedtls_ssl_session_save()` . When we use a ticket, `ticket_lifetime` is not valid.

ronald-cron-arm replied:
`ssl_ticket.c` does have `key->lifetime` the lifetime of the key with which the ticket was encrypted. This is also the lifetime of the ticket. Thus `ssl_ticket.c` have the ticket create time, its lifetime and can check it.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -1226,7 +1226,22 @@ struct mbedtls_ssl_session {
+     *  session, the ticket creation time may need to be saved for the ticket
+     *  module to be able to check the ticket age when the ticket is used.
+     *  That's the purpose of this field.
+     *  Before to create a new ticket, an Mbed TLS server set this field with
```
Minor: before creating a new ticket

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1223,8 +1224,28 @@ struct mbedtls_ssl_session {
     uint32_t MBEDTLS_PRIVATE(ticket_lifetime);   /*!< ticket lifetime hint    */
 #endif /* MBEDTLS_SSL_SESSION_TICKETS && MBEDTLS_SSL_CLI_C */
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) && defined(MBEDTLS_SSL_SRV_C) && \
+    defined(MBEDTLS_HAVE_TIME)
+    /*! When a ticket is created by a TLS server as part of an established TLS
+     *  session, the ticket creation time may need to be saved for the ticket
+     *  module to be able to check the ticket age when the ticket is used.
+     *  That's the purpose of this field.
+     *  Before to create a new ticket, an Mbed TLS server set this field with
+     *  its current time in milliseconds. This time may then be saved in the
+     *  session ticket data by the session ticket writing function and
+     *  recovered by the ticket parsing function later when the ticket is used.
+     *  The ticket module may then use this time to compute the ticket age and
+     *  determine if it has expired or not.
+     *  The Mbed TLS implementations of the session ticket writing and parsing
+     *  functions save and retrieve the ticket creation time as part of the
+     *  session ticket data. The session ticket parsing function relies on
+     *  the mbedtls_ssl_session_get_ticket_creation_time() API to get the
+     *  ticket creation time from the session ticket data.
+     */
+    mbedtls_ms_time_t MBEDTLS_PRIVATE(ticket_creation_time);
+#endif
```
`ticket_reception_time` is defined only in TLS 1.3 case thus not possible to reorganize things that way.

ronald-cron-arm commented on library/ssl_tls.c:
```diff
@@ -8908,6 +8909,7 @@ unsigned int mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg(
  *    opaque peer_cert<0..2^24-1>;    // length 0 means no peer cert
  *    opaque ticket<0..2^24-1>;       // length 0 means no ticket
  *    uint32 ticket_lifetime;
+ *    uint64 ticket_creation_time;
```
`start_time` may be useful as part of context save/load thus I prefer to keep it.

ronald-cron-arm commented on library/ssl_tls.c:
```diff
@@ -8900,6 +8900,7 @@ unsigned int mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg(
  *
  * struct {
  *    uint64 start_time;
+ *    uint8 endpoint;
```
Done for endpoint and ciphersuite.

ronald-cron-arm commented:
I've rebased and addressed the comments.

gilles-peskine-arm reviewed (changes requested):
This mostly looks good to me, with a doubt about partly preexisting design and a few minor change requests.

ronald-cron-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1227,8 +1228,28 @@ struct mbedtls_ssl_session {
+     *  session, the ticket creation time may need to be saved for the ticket
+     *  module to be able to check the ticket age when the ticket is used.
+     *  That's the purpose of this field.
+     *  Before creating a new ticket, an Mbed TLS server set this field with
+     *  its current time in milliseconds. This time may then be saved in the
+     *  session ticket data by the session ticket writing function and
+     *  recovered by the ticket parsing function later when the ticket is used.
```
First, thanks for the review.

>Info: why do it this way, rather than letting the ticket writing function query the current time if it wants?

The main reason is for an implementation of `mbedtls_ssl_ticket_write_t` to rely only on `mbedtls_ssl_session_save` to create the ticket session data and not having to set-up session field by itself (`ticket_creation_time` here). That way a third party implementation of ` mbedtls_ssl_ticket_write_t` handling `session` only as a black box with `mbedtls_ssl_session_save` should continue to work.

ronald-cron-arm commented:
@gilles-peskine-arm the CI is green (but ABI-API check as expected) and I believe I have addressed your comment now, please have another look, thanks.

gilles-peskine-arm reviewed (approved):
LGTM

tom-cosgrove-arm commented:
@mpg doing second review on this

mpg reviewed (approved):
LGTM!

mpg commented:
> partially supersede #8525

Was it really partial? Or can #8525 be closed now?

ronald-cron-arm commented:
> > partially supersede #8525
> 
> Was it really partial? Or can #8525 be closed now?

I have to check but yes I think there is still some work to do.
